File: dijkstra/src/routing.js
let R = require('ramda');

function newVertex(vertices, e) {
  return (
    (vertices.has(e.vertices[0]) && !vertices.has(e.vertices[1])) ||
    (vertices.has(e.vertices[1]) && !vertices.has(e.vertices[0]))
  );
}

function dijkstra(graph) {
  let vertices = new Map([[graph.vertices[0].id, { cost: 0, path: [] }]]);

  while (vertices.size < graph.vertices.length) {
    let start;
    let edge;
    let end;

    let cost = Number.MAX_VALUE;

    graph.edges.forEach((e) => {
      if (newVertex(vertices, e)) {
        if (
          vertices.has(e.vertices[0]) &&
          vertices.get(e.vertices[0]).cost + e.weight < cost
        ) {
          [start, end] = e.vertices;
          edge = e;
          cost = vertices.get(e.vertices[0]).cost + e.weight;
        } else if (
          vertices.has(e.vertices[1]) &&
          vertices.get(e.vertices[1]).cost + e.weight < cost
        ) {
          [end, start] = e.vertices;
          edge = e;
          cost = vertices.get(e.vertices[1]).cost + e.weight;
        }
      }
    });

    vertices.set(end, { cost, path: [...vertices.get(start).path, edge] });
  }

  return {
    cost: Array.from(vertices.values())
      .map((p) => p.cost)
      .reduce(R.add),
    vertices,
  };
}

function prim(graph) {
  let edges = R.range(0, graph.edges.length).map(R.F);
  let vertices = new Set([graph.vertices[0].id]);
  let totalCost = 0;

  while (vertices.size < graph.vertices.length) {
    let index;
    let cost = Number.MAX_VALUE;

    graph.edges.forEach((e, idx) => {
      if (newVertex(vertices, e) && e.weight < cost) {
        index = idx;
        cost = e.weight;
      }
    });

    edges[index] = true;
    totalCost += cost;

    let visit = vertices.has(graph.edges[index].vertices[0])
      ? graph.edges[index].vertices[1]
      : graph.edges[index].vertices[0];

    vertices.add(visit);
  }

  return {
    cost: totalCost,
    edges: graph.edges.filter((_, idx) => edges[idx]),
  };
}

module.exports = {
  dijkstra,
  prim,
};

File: dijkstra/routing.go
package main

import (
	"math"
)

func prim(graph []Node, start int) []Edge {
	path := make([]Edge, 0, len(graph)-1)

	visited := make(map[int]*Node, len(graph))
	visited[graph[start].id] = &graph[start]

	max := math.MaxInt
	longest := Edge{&graph[start], &max}
	for len(path) != len(graph)-1 {
		shortest := longest
		for _, n := range visited { // some redundancy...
			for _, e := range n.edges {
				if _, ok := visited[e.neig.id]; !ok && *e.weight < *shortest.weight {
					shortest = e
				}
			}
		}
		path = append(path, shortest)
		visited[shortest.neig.id] = shortest.neig
	}

	return path
}

func dijkstra(graph []Node, start int) []Edge {
	path := make([]Edge, 0, len(graph)-1)

	type cost struct {
		cost int
		node *Node
	}

	costs := make(map[int]cost, len(graph))
	costs[graph[start].id] = cost{0, &graph[start]}

	for len(path) != len(graph)-1 {
		lowCost := math.MaxInt
		var edge Edge

		for _, c := range costs { // some redundancy...
			for _, e := range c.node.edges {
				if _, ok := costs[e.neig.id]; ok {
					continue
				}
				cost := c.cost + *e.weight
				if cost < lowCost {
					lowCost = cost
					edge = e
				}
			}
		}

		path = append(path, edge)
		costs[edge.neig.id] = cost{lowCost, edge.neig}
	}

	return path
}

func cost(path []Edge) int {
	c := 0
	for _, e := range path {
		c = c + *e.weight
	}
	return c
}
